### ISSUE TITLE
LinkNet implementation working?

### ISSUE THREAD

**Issue by peteflorence** [meetshah1995/pytorch-semseg#16]
Hello, was excited to see you reimplemented LinkNet in PyTorch.

Can you verify whether the model has been tested?  I ran into a few issues, including that `linknet` is not included in the `get_model()` function in [models/__init__.py](https://github.com/meetshah1995/pytorch-semseg/blob/master/ptsemseg/models/__init__.py) and also this line in [`model/utils.py`](https://github.com/meetshah1995/pytorch-semseg/blob/master/ptsemseg/models/utils.py#L235) is broken. I seem to have fixed these two but am running into Cuda bad params world.  I wanted to ask first whether or not it's been tested (and if you could put me to a script that works with it) before diving in deeper.  Thanks!

**Comment by meetshah1995** [comment]
I haven't verified the linknet module myself. The changes you see were made by @ibadami. I don't plan to verify them in the coming weeks. The fix takes minimal effort, please feel free to send in a PR !

**Comment by peteflorence** [comment]
OK, thanks for letting me know.  Respectfully I think this issue should remain open though, because it is unknown whether or not the implementation is working.

**Comment by ibadami** [comment]
@peteflorence remove the 'nn.' in front of it. Cuda bad params can occur if your input dimensions are not as expected. Most often this occurs at the last layer where the target shape and last conv layer has mismatched shape. Kindly print the shape of the tensor after every operation. You will be able to see the what the problem is.

**Comment by ibadami** [comment]
in linknet from line 35 to 38 names of decoders are incorrect. Change them from self.decoder4 4 times to decoder4, decoder3 decoder2, decoder1. Something I could spot right away.
I will check what else is required to be corrected. Stay tuned!

**Comment by peteflorence** [comment]
Sure @ibadami.  Have you tested the implementation?  I'd be happy to help make some changes / PR, just want to get a sense of how well it's been tested.

**Comment by peteflorence** [comment]
OK @ibadami, thanks! :)

**Comment by ibadami** [comment]
@peteflorence So I just checked, It seems that the initial pspnet was copied from linknet while @meetshah1995 was working on the repo. I have changed the pspnet into fully functional linknet code. 
It is not really pspnet yet as there are few more changes required. But what I can assure you is that current pspnet is the corrected linknet.  Moreover there you can also find multiple resnet support which I added for pspnet. So I would suggest you to use pspnet from resnet encoder branch. Let me know if that code is throwing you any error. I will fix it as soon as possible. I know it is bit confusing. But it is work in progress.
Cheers!

**Comment by ibadami** [comment]
@peteflorence  checkout patch #17.

**Comment by peteflorence** [comment]
Thanks @ibadami, looking at it now

**Comment by peteflorence** [comment]
Hi @ibadami, I'm getting an error on `d4 = d4 + e3` in `forward()` in the model

When I add prints as below in `linknet.py`, I get these sizes for `d4` and `e3` respectively:
```
(1L, 256L, 24L, 30L)
(1L, 256L, 23L, 30L)
```

```
def forward(self, x):
        # Encoder
        x = self.convbnrelu1(x)
        x = self.maxpool(x)

        e1 = self.encoder1(x)
        e2 = self.encoder2(e1)
        e3 = self.encoder3(e2)
        e4 = self.encoder4(e3)

        # Decoder with Skip Connections
        d4 = self.decoder4(e4)
        print d4.size()
        print e3.size()
        d4 = d4 + e3
        d3 = self.decoder3(d4)
        d3 = d3 + e2
        d2 = self.decoder2(d3)
        d2 = d2 + e1
        d1 = self.decoder1(d2)

        # Final Classification
        f1 = self.finaldeconvbnrelu1(d1)
        f2 = self.finalconvbnrelu2(f1)
        f3 = self.finalconv3(f2)

        return f3
```

**Comment by peteflorence** [comment]
Hmm... maybe just need to crop/pad 360 since not divisible by 16?

**Comment by meetshah1995** [comment]
@peteflorence  - Yes, you are right. All datasets are at the moment not compatible with all the models due to differences in input sizes. 

My first task is to make all models working on Pascal VOC (512*512) and then add dataset specific cropping or padding arguments.

**Comment by peteflorence** [comment]
To close the loop here -- we have the training scripts 'running', but have not been able to get test segmentations that look meaningful.

If interested you can review my changes [here](https://github.com/peteflorence/pytorch-semseg/tree/pf-linknet) (just use `python train.py`, I changed the defaults in `train.py`)

**Comment by meetshah1995** [comment]
Thanks for the update. You haven't used the ResNet pretrained weights as initialization for the LinkNet encoder weights. Training from scratch did not get me good results with FCN-8s on Pascal VOC and I'd expect the same to happen with CamVid and LinkNet.

What would be ideal would be to have pre-trained ResNet from `torchvision` and use them to construct the linkNet as done [here](https://github.com/e-lab/LinkNet/blob/master/models/model-res-dec.lua#L30-L79). I have this in mind, but haven't quite found the time to implement them.

**Comment by ibadami** [comment]
@peteflorence @meetshah1995 , I just updated the linknet with pretrained resnet option.
Could you guys check it out?

**Comment by stereomatchingkiss** [comment]
I implement LinkNet at [here](https://github.com/stereomatchingkiss/blogCodes2/tree/master/python_deep_learning/segmentation/models), without using any pretrained weights, I am able to get mean accuracy close to 90%, you can see more details in this [post](http://forums.fast.ai/t/questions-on-semantic-segmentation/4853/6?u=tham).
